src/software/new_geom/util/almost_equal.cpp

```diff
@@ -0,0 +1,29 @@
+#include "software/new_geom/util/almost_equal.h"
+
+#include <algorithm>
+#include <cmath>
+
+#include "software/new_geom/geom_constants.h"
+
+// See
+// https://randomascii.wordpress.com/2012/02/25/comparing-floating-point-numbers-2012-edition/
+// for detailed explanation
+bool almostEqual(double a, double b)
```

Comment:
So I ran into an issue with the `EPSILON = 1e-15` value where for `bool collinear(const Point &a, const Point &b, const Point &c)` with inputs `a = (7.333333333333333, 2)`, `b = (5, -5)` and `c = (8, 4)` would return `false` (test file `intersects_test.cpp`, test case `ray_intersects_segment_edge`). Although they should have been collinear, the resulting cross product in `collinear` calculation was `2.6645352591003757e-15` which is greater than`EPSILON`.

To fix this problem, and dealing with the `double` precision issue in general, I followed a guide (https://randomascii.wordpress.com/2012/02/25/comparing-floating-point-numbers-2012-edition/) to implement a relative double comparison.

Then, using this new `almost_equal` function, I changed this line from `collinear` function from 1 to 2: 
1. `return std::fabs((b - a).cross(c - a)) < GeomConstants::EPSILON;`
2. `return almostEqual(v1.x() * v2.y(), v1.y() * v2.x());`

And since the `EPSILON` scales with the two `double`s being compared, it fixed the precision issue. 

Please let me know your thoughts on this!

Reply:
This seems good, I would add tests for these "barely collinear" cases though. Maybe I missed them but I didn't see a collinear_test with these

Reply:
Added the test case

Reply:
@garethellis0 see above for why I wrote the `almostEqual` function. A few other options I saw for resolving the double precision issue:
1. Increase the size of `EPSILON`, but this seems just as arbitrary
2. Change the interface of the relative comparison function `almostEqual` to pass in the `EPSILON` value
3. Define the `almostEqual` function in terms of ULP

Reply:
After doing a bit of reading on this, I'm OK with doing it if we make `eps` an argument to this function and do not give it a default value, because the user should be forced to choose the correct epsilon for their application, otherwise you can always come up with a sequence of floating point operations that accumulate enough error to violate your bound.

Also the implementation for this should definitely have a link to where we found this. 

Also if we're doing this, I think I prefer the solution shown under "No Silver Bullets" [here](https://bitbashing.io/comparing-floats.html) because it actually exposes all the implicit levers required for floating point comparison.

Reply:
Overall this is a good idea I think. We should probably get rid of all the `EPSILON` values in the codebase and replace them with specific absolute and UPS epsilons for specific use cases. It's a bit more of a pain to use, but hey, doing floating point comparisons correctly for geom is super important.

Reply:
tl;dr I agree EPS is arbitrary so we should be forced to specify it.

EPS can be different in many geometric contexts. In some cases we may compare points for equality by checking the distance between them. We could choose a different EPS value because our geometry is always used in meters, so a larger EPS value of 1e-9 (1 nano-meter) is plenty in this context.

In other cases we may be comparing the subtraction of cross products like what @jsn0716 ran into. Although this ultimately represents parallelism in some way, it's not as direct as the distance between points in 2D space, so a different EPS is appropriate here (one that may map back to some angular different that we deem acceptable).

Just some examples I guess. Perhaps we could specify these few "types" of EPS values in `geom_constants`?

Reply:
Changed `almostEqual` to use ULPs to compare doubles, and added the link to the explanation in header files. I defined a couple of the ULPs epsilon values just as `10` and `100` just so we can pick the appropriate value for an abitrary set of operations.